File: context/ai/induction/sdk/bun-server/services.ts
import { Profile } from "./models";
import { join } from "path";

export interface ProfileFetcher {
  fetch(profileId: string): Promise<Profile | undefined>;
}

export class HttpProfileFetcher implements ProfileFetcher {
  constructor(private serverUrl: string) {}

  async fetch(profileId: string): Promise<Profile | undefined> {
    try {
      const response = await fetch(`${this.serverUrl}/profiles/${profileId}`);
      if (response.ok) {
        return (await response.json()) as Profile;
      }
    } catch (e) {
      console.error(`[HttpProfileFetcher] Error fetching profile ${profileId}:`, e);
    }
    return undefined;
  }
}

export class FileProfileFetcher implements ProfileFetcher {
  constructor(private basePath: string) {}

  async fetch(profileId: string): Promise<Profile | undefined> {
    const filePath = join(this.basePath, `${profileId}.json`);
    const file = Bun.file(filePath);
    if (await file.exists()) {
      return (await file.json()) as Profile;
    }
    return undefined;
  }
}

export class CompositeProfileFetcher implements ProfileFetcher {
  constructor(private fetchers: ProfileFetcher[]) {}

  async fetch(profileId: string): Promise<Profile | undefined> {
    for (const fetcher of this.fetchers) {
      const profile = await fetcher.fetch(profileId);
      if (profile) {
        return profile;
      }
    }
    return undefined;
  }
}

export const Environment = {
  isProduction: () => process.env.APP_ENV === "production" || process.env.APP_ENV === "prod",
};

File: context/ai/induction/sdk/src/main/scala/com/induction/sdk/services.scala
package com.induction.sdk

import sttp.client3.*
import upickle.default.*
import java.nio.file.{Files, Paths}

trait ProfileFetcher:
  def fetch(profileId: String): Option[Profile]

class HttpProfileFetcher(serverUrl: String) extends ProfileFetcher:
  private val backend = HttpClientSyncBackend()

  override def fetch(profileId: String): Option[Profile] =
    val request = basicRequest
      .get(uri"$serverUrl/profiles/$profileId")
      .response(asStringAlways)

    val response = request.send(backend)
    if response.code.isSuccess then
      Some(read[Profile](response.body))
    else
      None

class FileProfileFetcher(basePath: String) extends ProfileFetcher:
  override def fetch(profileId: String): Option[Profile] =
    val path = Paths.get(basePath, s"$profileId.json")
    if Files.exists(path) then
      val content = Files.readString(path)
      Some(read[Profile](content))
    else
      None

class CompositeProfileFetcher(fetchers: List[ProfileFetcher]) extends ProfileFetcher:
  override def fetch(profileId: String): Option[Profile] =
    fetchers.view.flatMap(_.fetch(profileId)).headOption

object Environment:
  def isProduction: Boolean =
    sys.env.get("APP_ENV").contains("production")
